### ISSUE TITLE
[ARM CI]Introduce ARM CI to bucardo community

### ISSUE THREAD

**Issue by bzhaoopenstack** [bucardo/bucardo#221]
Currently, the CI of bucardo(in github https://github.com/bucardo/bucardo/) is handled by Travis-CI .  But most jobs are running under X86 hw ARCH. This leads an problem that we don't have a good way to test every pull request that the proposal can be deployed, compiled or tested success on ARM.

We should add a CI system that support ARM ARCH for more resources and less human resource invest. Using it, Bucardo can officially support arm release in the future. 

We have 2 alternatives:
1. Using existing Travis-CI to test on aarch64, but the performance of aarch64 machines in Travis seems not very good, and lack so many build tools on aarch64 platform.
2. OpenLab

Here I'd like to introduce OpenLab to the community. [OpenLab |https://openlabtesting.org/] is a open source CI system that can test any open source software on either x86 or arm ARCH, it's mainly used by github projects. Now some [projects|https://github.com/theopenlab/openlab-zuul-jobs/blob/master/zuul.d/jobs.yaml] has integrated it already. Such as containerd (a CNCF project),  terraform and so on.

OpenLab uses the open source CI software [Zuul |https://github.com/openstack-infra/zuul] for  CI system. Zuul is used by OpenStack community as well. integrating with OpenLab is quite easy using its github app. All config info is open source as well. 

If Bucardo team has interested with it, I can help for the integration. Thank you.

**Comment by bzhaoopenstack** [comment]
Hi team,
@turnstep @jonjensen  @machack666, could you please give some ideas about ARM CI? For now, I got all pass on ubuntu1804 aarch64 platform..

**Comment by jonjensen** [comment]
@bzhaoopenstack That seems worthwhile to me. Thanks for testing it already.

What do we need to do to make the tests run after each commit?

Also, FYI, I get a TLS certificate error going to https://openlabtesting.org/ because the cert expired 3 months ago.

**Comment by bzhaoopenstack** [comment]
@jonjensen Thanks very much for the info about the TLS error. That must be my fault in previous exchange the new TLS cert last month.. I had already updated it. And now, the website of openlab is available.

Hmm, let me think, we have 2 options for the each commit check towards aarch64 platform.
1. The first one is using exist Travis CI that I already test on ubuntu bionic aarch64, but seems the ubuntu arm repo in travis build-in image only has PG-10 packages, also for ubuntu 1604 aarch64 travis image, that image lacks so many aarch64 deps, so  I can only get pass on travis ubuntu bionic with test params PG=10 and perl=5.24(https://travis-ci.com/github/bzhaoopenstack/bucardo/jobs/434139481). There is no other PG version providered, and there are some issues about `perlbrew` tools on travis ubuntu bionic. So the all status looks (https://travis-ci.com/github/bzhaoopenstack/bucardo/builds/200198633). If bucardo team wants this, seems we can only add 1 couple(PG=10, perl=5.24) and use bionic image into travis configuration file.
2. Using openlab, community needs to install a openlab github app. Then it will run like this(https://github.com/kubernetes/cloud-provider-openstack/pull/1311), the `theopenlab-ci` robot will leave the test comment in PR for each commit. But before that, we need to config something in openlab, also in `bucardo` repo like this(https://github.com/kubernetes/cloud-provider-openstack/blob/master/.zuul.yaml) to config which job need to be run.

**Comment by jonjensen** [comment]
Hmm. We don't know of anyone using ARM for Bucardo, and Postgres 10 + Perl 5.24 are both fairly dated now too.

We have limited expert help maintaining Bucardo, so I'm going to say we should leave this on the back-burner for now, and we can reconsider later. At the moment we have a higher need to work on our documentation and some new features for the next release.

@bzhaoopenstack Thanks for opening this issue, and hopefully we'll be able to consider it later.

**Comment by bzhaoopenstack** [comment]
@jonjensen Thanks for explain more and aware.

Recently, I use the master branch of bucardo in a mixed cluster for a demo show recently, which means we have 2 parts machines, one part is X86 machines, the other is ARM machines. And we get a request from customer, they wants to migrate the existing X86 cluster to ARM clusters one by one due to they can not exchange the whole X86 machines to ARM directly, so there will be a mid-progress scenario, that's 2 parts of machines in customer's DC, one is X86, the other is ARM. In the end, all X86 will be exchanged by ARM machines. That's the reason we consider bucardo needs support accross hardware platform. And it does, I tried the demo, and bucardo works well on both ARCH.

Hmm, like my first description in this PR, that is `"If Bucardo team has interested with it, I can help for the integration. Thank you."` If Bucardo team has no more bandwith, I think I can help to integrate and maintain the CI works well in the future. But the first step is Bucardo team(you guys) need to show me a direction, such as which integration way is accepted by you, travis? or openlab? .. Then we can start to do this and make ARM CI better steps by steps. How do you think? ;-)

Thank you

**Comment by bzhaoopenstack** [comment]
Hi guys, I opened a PR for this, due to the previous reason, I test so many times, seems that would cost the least time to run on travis ARM machines... And I set the [WIP] title when we are all ready about that. Thank you.

**Comment by bzhaoopenstack** [comment]
Oh, @jonjensen Did team abandon travis? Due to they become taking money? Hmm, not sure why the PRhttps://github.com/bucardo/bucardo/pull/223 doen't run on travis or any test platform on Bucardo master branch.

**Comment by bzhaoopenstack** [comment]
Hi team,
@turnstep @jonjensen @machack666 Could you please take a look at [PR](https://github.com/bucardo/bucardo/pull/223) ? And give some advices and feedback. Thank you very much

**Comment by jonjensen** [comment]
@bzhaoopenstack I don't know what is going on in the Travis CI world with open source, but if it ends up being a hassle like I'm reading in [this thread](https://twitter.com/alex_gaynor/status/1336098972864303104) we should probably just forget about it. Any other ARM CI you'd like to try that is free for open source projects?

**Comment by bzhaoopenstack** [comment]
@jonjensen Thanks for feedback.

Hmm, I suggested 2 options in [previous comment](https://github.com/bucardo/bucardo/issues/221#issuecomment-725789229) , so if bucardo team plan to abandon travis-ci. As much I know, that we can use openlab ci(free for ARM and X86 test resources), which mentioned the option 2 would be good. But before that, we need to know the decision of bucardo team about whether abandon travis-ci and whether accept openlab CI.

**Comment by jonjensen** [comment]
@bzhaoopenstack Thanks for checking back. When it comes to CI, I think all of us on the Bucardo team are happy to have any volunteer help do it however they want, if it doesn't make extra work for us to deal with! If you want to set up CI there, and there's no cost, please go ahead. If it works well, that will probably become the new official Bucardo CI. I will be happy to take a look at what you do, but can't offer more time than that myself right now. Thanks!

**Comment by bzhaoopenstack** [comment]
Hi @jonjensen 
So sorry for late reply, so busy recently.
Thanks. I will start the work and test on the test env first. And back to get your review for the test job definition. Once the whole pre-work is done, I will request help from you guys to config bucardo repo in github and how to use. Thanks. ;-)

**Comment by jonjensen** [comment]
@bzhaoopenstack Sounds great. Thanks!

**Comment by bzhaoopenstack** [comment]
@jonjensen Thanks, I had finished the configuration towards my forked bucardo repo, and in test progress, once I finish, will let you know and review.

**Comment by bzhaoopenstack** [comment]
@jonjensen 
It almost done here, https://github.com/bzhaoopenstack/bucardo/pull/4#issuecomment-768079740 ..
Please see whether it can match our requirement, right now, introduce 4 jobs to each bucardo PR, 3 jobs run on x86 with PG 10,11,12 and perl 5.30 . 1 job runs on arm64 with PG 12(upstream repo only provides arm64 packages after PG12) and perl 5.30.

How do you think? If you are OK with the test demo, we can push it here into bucardo official repo.

**Comment by bzhaoopenstack** [comment]
But hit issues which is not related with the CI system. So I have no idea to resolve now

**Comment by bzhaoopenstack** [comment]
Hi @jonjensen any updates or suggests for the demo, there is no steps for a month..

**Comment by jonjensen** [comment]
@bzhaoopenstack It sure seems that the problem must have something to do with the CI system, because the unit tests all pass fine on my machine (Fedora 33 x86_64):

```
% make test
Skip blib/lib/Bucardo.pm (unchanged)
cp bucardo blib/script/bucardo
"/home/testuser/.plenv/versions/5.32.0/bin/perl5.32.0" -MExtUtils::MY -e 'MY->fixin(shift)' -- blib/script/bucardo
PERL_DL_NONLAZY=1 "/home/testuser/.plenv/versions/5.32.0/bin/perl5.32.0" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/00-release.t ............... skipped: Test skipped unless environment variable RELEASE_TESTING is set
t/01-basic.t ................. ok
t/02-bctl-customname.t ....... ok
t/02-bctl-db.t ............... ok
t/02-bctl-dbg.t .............. ok
t/02-bctl-herd.t ............. ok
t/02-bctl-sync.t ............. ok
t/02-bctl-table.t ............ ok
t/10-fullcopy.t .............. ok
t/10-makedelta.t ............. 2/? WARNING:  PID 25681 is not a PostgreSQL server process
t/10-makedelta.t ............. ok
t/10-object-names.t .......... ok
t/20-drizzle.t ............... skipped: Cannot test Drizzle unless the Perl module DBD::drizzle is installed
t/20-firebird.t .............. skipped: Cannot test Firebird unless the Perl module DBD::Firebird is installed
t/20-mariadb.t ............... skipped: Cannot test MariaDB unless the Perl module DBD::mysql is installed
t/20-mongo.t ................. skipped: Cannot test mongo unless the Perl module MongoDB is installed
t/20-mysql.t ................. skipped: Cannot test MySQL unless the Perl module DBD::mysql is installed
t/20-oracle.t ................ skipped: Cannot test Oracle unless the Perl module DBD::Oracle is installed
t/20-postgres.t .............. ok
t/20-redis.t ................. skipped: Cannot test Redis unless the Perl module Redis is installed
t/20-sqlite.t ................ skipped: Cannot test SQLite unless the Perl module DBD::SQLite is installed
t/30-crash.t ................. ok
t/30-delta.t ................. ok
t/40-conflict.t .............. ok
t/40-customcode-exception.t .. ok
t/40-serializable.t .......... ok
t/50-star.t .................. ok
t/98-cleanup.t ............... ok
t/99-lint.t .................. skipped: Test skipped unless environment variable RELEASE_TESTING is set
t/99-perlcritic.t ............ skipped: Test skipped unless environment variable RELEASE_TESTING is set
t/99-signature.t ............. skipped: Test skipped unless environment variable RELEASE_TESTING is set
t/99-spellcheck.t ............ skipped: Test skipped unless environment variable RELEASE_TESTING is set
t/99-yaml.t .................. skipped: Test skipped unless environment variable RELEASE_TESTING is set
All tests successful.
Files=32, Tests=1600, 413 wallclock secs ( 0.57 usr  0.10 sys + 113.88 cusr 32.17 csys = 146.72 CPU)
Result: PASS
```

**Comment by bzhaoopenstack** [comment]
https://github.com/bzhaoopenstack/bucardo/pull/4#issuecomment-797196563  Oh, yeah, I retest via the openlab, it get pass.

**Comment by jonjensen** [comment]
@bzhaoopenstack Excellent!

So now we can close PR #4 right?

And what do we need to do to make the OpenLab tests run for every PR? Is that happening now automatically?

**Comment by bzhaoopenstack** [comment]
@jonjensen we need to install the [openlabci app](https://github.com/apps/theopenlab-ci) in bucardo project, and after that, I will reconfig the openlab server to our official repo bucardo/bucardo and nochange other config. As currently, the CI had been configured as a PR triggered tests, the only difference is those jobs only can enable on my forked repo bzhaoopenstack/bucardo, which already installed the openlabci app. Then create a PR to post into bucardo/bucardo repo like https://github.com/bzhaoopenstack/bucardo/blob/master/.zuul.yaml  as a openlab zuul config. That's all

**Comment by bzhaoopenstack** [comment]
For installing the openlabci app, that should be operated by the admin of bucardo/bucardo repo

**Comment by bzhaoopenstack** [comment]
I think we can step up the related work recently, right? @jonjensen

**Comment by amitdkhan-pg** [comment]
Having a PR-triggered openlab CI system looks really beneficial for the bucardo project as a whole, since there does not seem to be any Travis or Openlab CI running currently. And it looks like @bzhaoopenstack is ready to independently do all the setup required to bring the openlab CI system up and running. He has already done the necessary changes in https://github.com/bzhaoopenstack/bucardo/commit/7c10b0fb8fba53eeb747218782010848f949e275

The only thing @bzhaoopenstack can't do by himself is to install the openlabci app. I understand it all depends on the policies made by the bucardo project owners.

@jonjensen wrote :
> And what do we need to do to make the OpenLab tests run for every PR? Is that happening now automatically?

@jonjensen , is it possible to get us in touch with the bucardo project admin to discuss if it's feasible to install the openlabci app using bucardo account ? Like I mentioned above, this is the only step that @bzhaoopenstack or me cannot do ourselves. @bzhaoopenstack has taken care of all the other steps, I believe.
